test/garbage-test.js

```diff
@@ -0,0 +1,92 @@
+import { describe, it } from 'mocha'
+import assert from 'assert'
+
+import { resolve, reject, fulfill, never } from '../src/main'
+import getCounts, { collect, formatResults, sumResults } from './lib/refcount-util'
+
+describe('reference counts', () => {
```

Comment:
You might want to have a look at the commented "*should be interesting*" test cases. Uncomment them and run them, maybe even replace that `sumResults` callback with `log`.

The `refcount` is based on what a garbage collector would do to see which objects have to be retained in memory. Certain "root" objects are unclaimable: the global scope, all scopes currently on the stack, and any objects (typically callbacks) that ongoing asynchronous operations cling onto. It then has to keep everything that is referenced by them. The `collect` method does quite this - you pass in `sources`, and it counts all the objects referenced by them, giving an overview by constructor name (which you can filter with `targets`). To avoid counting multiple times, it keeps a `WeakSet`. (Actually now that I'm rethinking it, one could just have put everything into a `Set` first and then iterated that for the count). Because creed uses objects everywhere instead of callbacks, we can inspect everything 
In a promise library, the sources are typically a) the promise that marks the end of the chain, which could be stored somewhere (for example in a cache). This is what `getCounts` stores as `result`.
And there are the resolver callbacks of running async operations referencing the promises they resolve. This is were `monitoredDelay` comes in, which can be used to simulate an async action (taking a fixed 5ms). A `getCount` call allows to simulate one run of a promise chain - it returns a future that is resolved when no more delays are opened. The `monitoredTimeout` keeps track of these currently ongoing actions, adding the values which the "closure" captures (the delay promise and its result) to the `globals` set and removing them when the timeout is done.
In one run of `getCounts`, right before and after every delay is fulfilled, as well as after the initial execution, counts of objects are `collect`ed with either `result`, `globals` or the current `captures` for the source, then passed to the `report` callback for analisation and stored in `res`ults. For entertainment(?), `collect` does not only count how many objects of which type it sees, but also how many of them have already been encountered before during this run and how many are new since the last collection.

I hope the recursions to build the respective "chains" should be self-explaining. Use the `log` to get an intuition for the objects being created over time. The assertions test the counts not to get over some threshold.